numba/cuda/compiler.py

```diff
@@ -1123,6 +1026,40 @@ def compile(self, sig):
             self.sigs.append(sig)
         return kernel
 
+    def compile_device(self, args):
+        """Compile the function for the given argument types.
+
+        Each signature is compiled once by caching the compiled function inside
+        this object.
+
+        Returns the `CompileResult`.
+        """
+        if args not in self.overloads:
```

Comment:
What is an overload was compiled for the `args`, but that compilation produced a kernel and not a device function?

Reply:
Somewhere near to here, we need to enforce casting (or failure to compile) if a signature was provided.